Source/WebCore/platform/ScrollSnapAnimatorState.cpp

```diff
@@ -95,18 +95,72 @@ float ScrollSnapAnimatorState::adjustedScrollDestination(ScrollEventAxis axis, F
+    }
+}
+
+void ScrollSnapAnimatorState::preserveCurrentTargetForAxis(ScrollEventAxis axis)
+{
+    auto snapID = activeSnapIDForAxis(axis);
+    auto snapOffsets = snapOffsetsForAxis(axis);
+    
+    auto found = std::find_if(snapOffsets.begin(), snapOffsets.end(), [snapID](SnapOffset<LayoutUnit> p) -> bool {
+        return p.snapTargetId == *snapID;
+    });
+    if (found != snapOffsets.end())
+        setActiveSnapIndexForAxis(axis, std::distance(snapOffsets.begin(), found));
+}
+
 bool ScrollSnapAnimatorState::resnapAfterLayout(ScrollOffset scrollOffset, const ScrollExtents& scrollExtents, float pageScale)
 {
+    // Check if we need to set the active target to a focused element
+    setFocusedElementForAxis(ScrollEventAxis::Vertical);
+    setFocusedElementForAxis(ScrollEventAxis::Horizontal);
+
     bool snapPointChanged = false;
-    // If we are already snapped in a particular axis, maintain that. Otherwise, snap to the nearest eligible snap point.
     auto activeHorizontalIndex = activeSnapIndexForAxis(ScrollEventAxis::Horizontal);
-    if (!activeHorizontalIndex || *activeHorizontalIndex >= snapOffsetsForAxis(ScrollEventAxis::Horizontal).size())
-        snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Horizontal, scrollOffset, scrollExtents, pageScale);
-
     auto activeVerticalIndex = activeSnapIndexForAxis(ScrollEventAxis::Vertical);
-    if (!activeVerticalIndex || *activeVerticalIndex >= snapOffsetsForAxis(ScrollEventAxis::Vertical).size())
+    auto activeHorizontalID = activeSnapIDForAxis(ScrollEventAxis::Horizontal);
+    auto activeVerticalID = activeSnapIDForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsVertical = snapOffsetsForAxis(ScrollEventAxis::Vertical);
+    auto snapOffsetsHorizontal = snapOffsetsForAxis(ScrollEventAxis::Horizontal);
+    
+    // Check if we need to set the current indeces
+    if (!activeVerticalIndex || *activeVerticalIndex >= snapOffsetsForAxis(ScrollEventAxis::Vertical).size()) {
         snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Vertical, scrollOffset, scrollExtents, pageScale);
-
+        activeVerticalIndex = activeSnapIndexForAxis(ScrollEventAxis::Vertical);
+    }
+    if (!activeHorizontalIndex || *activeHorizontalIndex >= snapOffsetsForAxis(ScrollEventAxis::Horizontal).size()) {
+        snapPointChanged |= setNearestScrollSnapIndexForAxisAndOffset(ScrollEventAxis::Horizontal, scrollOffset, scrollExtents, pageScale);
+        activeHorizontalIndex = activeSnapIndexForAxis(ScrollEventAxis::Horizontal);
+    }
+    
+    // If we have an active target, see if we need to preserve it
+    if (activeHorizontalID && activeHorizontalIndex) {
+        if (*activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
```

Comment:
If combined into one if statement, this would be a bit clearer, I think. For instance:

```c++
    if (activeHorizontalID && activeHorizontalIndex &&
        *activeHorizontalID != snapOffsetsHorizontal[*activeHorizontalIndex].snapTargetId)
```

Reply:
I think we want to leave the logic like this for horizontal axis. This is due to an implementation detail where we need to choose an axis to prioritize the target for (we choose horizontal). This prevents us from trying to follow the target of the vertical axis when we are already following the target of the horizontal axis.